subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/HealthExpirationStrategy.java

```diff
@@ -16,26 +16,115 @@
 
 package org.gradle.launcher.daemon.server.health;
 
-import com.google.common.collect.ImmutableList;
-import org.gradle.launcher.daemon.server.expiry.AnyDaemonExpirationStrategy;
+import com.google.common.base.Joiner;
+import org.gradle.api.internal.DocumentationRegistry;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.internal.util.NumberUtil;
 import org.gradle.launcher.daemon.server.expiry.DaemonExpirationResult;
+import org.gradle.launcher.daemon.server.expiry.DaemonExpirationStatus;
 import org.gradle.launcher.daemon.server.expiry.DaemonExpirationStrategy;
+import org.gradle.launcher.daemon.server.health.gc.GarbageCollectionStats;
+import org.gradle.launcher.daemon.server.health.gc.GarbageCollectorMonitoringStrategy;
 
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.gradle.launcher.daemon.server.expiry.DaemonExpirationStatus.DO_NOT_EXPIRE;
+import static org.gradle.launcher.daemon.server.expiry.DaemonExpirationStatus.GRACEFUL_EXPIRE;
+import static org.gradle.launcher.daemon.server.expiry.DaemonExpirationStatus.IMMEDIATE_EXPIRE;
+import static org.gradle.launcher.daemon.server.expiry.DaemonExpirationStatus.highestPriorityOf;
+
+/**
+ * A {@link DaemonExpirationStrategy} which monitors daemon health and expires the daemon
+ * whenever unhealthy conditions are detected. Currently, this strategy monitors JVM memory
+ * health by detecting GC thrashing and excessive heap or metaspace usage. In addition to
+ * expiring the daemon, whenever unhealthy conditions are detected, this strategy will
+ * print a warning log to the console informing the user of the issue and instructing them
+ * on how to adjust daemon memory settings.
+ */
 public class HealthExpirationStrategy implements DaemonExpirationStrategy {
 
-    private final DaemonExpirationStrategy strategy;
+    /**
+     * A system property which enables this strategy. Defaults to true.
+     */
+    public static final String ENABLE_PERFORMANCE_MONITORING = "org.gradle.daemon.performance.enable-monitoring";
+
+    /**
+     * A system property which disables logging upon expiration events. Defaults to false.
+     */
+    public static final String DISABLE_PERFORMANCE_LOGGING = "org.gradle.daemon.performance.disable-logging";
 
-    public HealthExpirationStrategy(DaemonMemoryStatus memoryStatus) {
-        this.strategy = new AnyDaemonExpirationStrategy(ImmutableList.of(
-            new GcThrashingDaemonExpirationStrategy(memoryStatus),
-            new LowHeapSpaceDaemonExpirationStrategy(memoryStatus),
-            new LowNonHeapDaemonExpirationStrategy(memoryStatus)
-        ));
```

Comment:
The previous implementation felt needlessly abstracted for whatever advantages we gained from the separation. If you excluded the logging that these separate strategies were doing (which would be moved out of each strategy even if we left them as-is otherwise), then we were effectively encapsulating a single if statement for each of `LowHeapSpaceDaemonExpirationStrategy`, `LowNonHeapDaemonExpirationStrategy`, and `GcThrashingDaemonExpirationStrategy`.

That IMO made reasoning about the code more tedious than it needed to be -- especially given that all of the now-consolidated expiration strategies all were very closely related in regards to the checks they were performing. This seemed like a no-brainer change for me which consolidated extremely related logic into a single file.

Reply:
Hmmm...that's a little surprising to me, but I also don't think it's a big deal, either.  I was mostly just curious.  If you feel like this is easier to reason about now, then it's fine.